### PR TITLE
test: add unit test for permissions

### LINKED ISSUE #809: test: add unit test for permissions
### Describe the feature you'd like

https://github.com/bucketeer-io/bucketeer/blob/4e36c0589b1ad514174a3dabfc499f0a2db36d42/pkg/autoops/api/api.go#L822

and others in this package

### Background

_No response_

### Alternative solutions

_No response_

### PR DESCRIPTION
**Pull request by kakcy**
This pull request primarily focuses on adding tests to verify that APIs are successful with a Viewer account that is not SystemAdmin. This is done across multiple files and services, including `pkg/account/api/api_key_test.go`, `pkg/auditlog/api/api_test.go`, `pkg/autoops/api/api_test.go`, `pkg/feature/api/api_test.go`, `pkg/feature/api/flag_trigger_test.go`, and `pkg/feature/api/segment_test.go`. 

resolve #809
Includes tests for the following services:
- feature
- autoops
- auditlog
- account

This pull request includes significant changes to the test files `api_key_test.go`, `api_test.go`, and `api_test.go` in the `pkg/account/api`, `pkg/auditlog/api`, and `pkg/autoops/api` directories respectively. The changes modify the structure of test cases and add new test scenarios to improve test coverage. The changes mainly involve adding a `service` and `context` field to the test patterns, modifying the `createError` function to take `localizer` as a parameter, and adding new test cases to handle viewer accounts.

Here are the top five most important changes:

1. **Changes to `api_key_test.go`**:
   - Added `service` and `context` fields to test patterns in `TestGetAPIKeyMySQL` and `TestListAPIKeysMySQL` functions. This allows each test case to use a specific service and context. [[1]](diffhunk://#diff-d172cb59f4c01d961a5b53ec2e7df81dad55a5a6a7c794a7b8d0759590dec032R448-R480) [[2]](diffhunk://#diff-d172cb59f4c01d961a5b53ec2e7df81dad55a5a6a7c794a7b8d0759590dec032R561-R652)
   - Modified the `createError` function to take `localizer` as a parameter. This allows each test case to use a specific localizer. [[1]](diffhunk://#diff-d172cb59f4c01d961a5b53ec2e7df81dad55a5a6a7c794a7b8d0759590dec032L437-R437) [[2]](diffhunk://#diff-d172cb59f4c01d961a5b53ec2e7df81dad55a5a6a7c794a7b8d0759590dec032L509-R550)
   - Added new test cases to handle viewer accounts. This improves test coverage by testing the behavior of the system when the user is a viewer. [[1]](diffhunk://#diff-d172cb59f4c01d961a5b53ec2e7df81dad55a5a6a7c794a7b8d0759590dec032L485-R537) [[2]](diffhunk://#diff-d172cb59f4c01d961a5b53ec2e7df81dad55a5a6a7c794a7b8d0759590dec032R561-R652)

2. **Changes to `api_test.go` in `pkg/auditlog/api`**:
   - Added `service` and `context` fields to test patterns in `TestListAuditLogsMySQL`, `TestListAdminAuditLogsMySQL`, and `TestListFeatureHistoryMySQL` functions. This allows each test case to use a specific service and context. [[1]](diffhunk://#diff-b52b70ada54cdb42bd516f91b21c5b72cb85999f9434892391a2ed031ac9b4feR70-R148) [[2]](diffhunk://#diff-b52b70ada54cdb42bd516f91b21c5b72cb85999f9434892391a2ed031ac9b4feR239-R275) [[3]](diffhunk://#diff-b52b70ada54cdb42bd516f91b21c5b72cb85999f9434892391a2ed031ac9b4feR352-R377)
   - Modified the `createError` function to take `localizer` as a parameter. This allows each test case to use a specific localizer. [[1]](diffhunk://#diff-b52b70ada54cdb42bd516f91b21c5b72cb85999f9434892391a2ed031ac9b4feL59-R59) [[2]](diffhunk://#diff-b52b70ada54cdb42bd516f91b21c5b72cb85999f9434892391a2ed031ac9b4feL197-R228)
   - Added new test cases to handle viewer accounts and unassigned role accounts. This improves test coverage by testing the behavior of the system when the user is a viewer or has an unassigned role. [[1]](diffhunk://#diff-b52b70ada54cdb42bd516f91b21c5b72cb85999f9434892391a2ed031ac9b4feR239-R275) [[2]](diffhunk://#diff-b52b70ada54cdb42bd516f91b21c5b72cb85999f9434892391a2ed031ac9b4feR352-R377)

3. **Changes to `api_test.go` in `pkg/autoops/api`**:
   - Added `service` field to test patterns in `TestGetAutoOpsRuleMySQL` function. This allows each test case to use a specific service.
   - Added new test case to handle viewer accounts. This improves test coverage by testing the behavior of the system when the user is a viewer.

4. **Changes to `createContextWithToken` function in `pkg/auditlog/api/api_test.go`**:
   - Added `isSystemAdmin` parameter to `createContextWithToken` function. This allows the creation of a context with a token where `IsSystemAdmin` can be set as needed.

5. **Changes to `newAuditLogService` function in `pkg/auditlog/api/api_test.go`**:
   - Added `newAuditLogServiceForViewer` function. This allows the creation of an audit log service for a viewer account.
Here are the top five most important changes:

1. `pkg/account/api/api_key_test.go`: A new test function `TestViewerEnvironmentRole` is added to test APIs with a Viewer account that is not SystemAdmin.

2. `pkg/auditlog/api/api_test.go`: Two new test functions `TestViewerEnvironmentRole` and `newAuditLogServiceForViewer` are added to test APIs with a Viewer account that is not SystemAdmin. [[1]](diffhunk://#diff-b52b70ada54cdb42bd516f91b21c5b72cb85999f9434892391a2ed031ac9b4feR263-R314) [[2]](diffhunk://#diff-b52b70ada54cdb42bd516f91b21c5b72cb85999f9434892391a2ed031ac9b4feR341-R366)

3. `pkg/autoops/api/api_test.go`: Two new test functions `TestViewerEnvironmentRole` and `createFeatureServiceForViewer` are added to test APIs with a Viewer account that is not SystemAdmin. [[1]](diffhunk://#diff-e706076973d3d7e75d0e604529c9a1491f34d5b26f86d7f56183ddd55c7f74b9R768-R843) [[2]](diffhunk://#diff-e706076973d3d7e75d0e604529c9a1491f34d5b26f86d7f56183ddd55c7f74b9R908-R931)

4. `pkg/feature/api/api_test.go`: Two new test functions `createFeatureServiceForViewer` and `TestViewerEnvironmentRoleForFeatureAPIs` are added to test APIs with a Viewer account that is not SystemAdmin. [[1]](diffhunk://#diff-b900941517f1d0f0d7f739282ddc32ac4cbdb33f4bd2291b1a00d4d7d10912dbR167-R199) [[2]](diffhunk://#diff-fad41f54a190634386c37788c6f063b9d12ddb75b1928675e5b6ddd61b0d9663R1477-R1624)

5. `pkg/feature/api/flag_trigger_test.go`: A new test function `TestPermissionForAcquisitionAPIForFlagTrigger` is added to test APIs with a Viewer account that is not SystemAdmin.

These changes are designed to improve the test coverage and ensure that the application behaves as expected when accessed by a Viewer account that is not SystemAdmin.

### CODE REVIEW

**Review by kentakozuka — COMMENTED** [review]
IMHO, these test patterns should be added to existing tests such as `TestEnableAPIKeyMySQL` or so. WDYT?

**Comment by kakcy** [comment]
I think it would be easier to read since it would be a single test function for the same API, so I would consider it.

**Comment by kakcy** [comment]
I tried it based on `TestListAdminAuditLogsMySQL()`, but it required a lot of changes from the existing test code, so I'd like to keep it as is.
The reason there are so many changes is because the context of this test is different from previous tests.
The existing test code is basically running with `IDToken.IsSystemAdmin` `true`, but
For our test, we need to test with `false` to test correctly.

This is because [CheckEnvironmentRole()](https://github.com/bucketeer-io/bucketeer/blob/main/pkg/role/role.go#L71) is determined based on the fixed authority of `AccountV2_Role_Environment_EDITOR` in the case of `SystemAdmin`.

If you change the Context for each test, you will also need to define a Localizer, and the expected errors will also need to be generated separately.
As a result, other test codes may become complicated.

There are tests that combine different APIs depending on the purpose, like the test below, so I would like to use the current situation this time. 
https://github.com/bucketeer-io/bucketeer/blob/main/pkg/feature/api/feature_test.go#L1243

What do you think?

**Comment by kakcy** [comment]
I reviewed the process and modified it to include it as part of the pattern for test functions related to the API.
https://github.com/bucketeer-io/bucketeer/pull/841/commits/71de46ee31ce88da54e0d85d351acf0c9df8ff2f

Please check whether the modifications are what you intended.
Also fix other tests if they match your intent.

**Comment by kentakozuka** [comment]
@kakcy Can you resolve conflicts?

**Comment by kakcy** [comment]
@kentakozuka I resolved Conflicts.

**Comment by kentakozuka** [comment]
@kakcy https://github.com/bucketeer-io/bucketeer/commit/71de46ee31ce88da54e0d85d351acf0c9df8ff2f looks good to me 👍 Thanks!

**Comment by kakcy** [comment]
Refactored permission tests to combine them into each API's test function.
Please check when you have time 🙏

File: pkg/account/api/api_key_test.go [hunk 1]
```diff
@@ -450,28 +444,40 @@ func TestGetAPIKeyMySQL(t *testing.T) {
 	}
 
 	patterns := []struct {
-		desc     string
-		setup    func(*AccountService)
-		req      *accountproto.GetAPIKeyRequest
-		expected error
+		desc           string
+		service        *AccountService
```
Comment (kentakozuka) [inline pkg/account/api/api_key_test.go:448]: We can create the service in the test loop. plz remove the same line in the other tests as well.

File: pkg/account/api/api_key_test.go [hunk 2]
```diff
@@ -481,19 +487,54 @@ func TestGetAPIKeyMySQL(t *testing.T) {
...
 			},
-			req:      &accountproto.GetAPIKeyRequest{Id: "id"},
-			expected: nil,
+			req: &accountproto.GetAPIKeyRequest{Id: "id"},
+			getExpectedErr: func(localizer locale.Localizer) error {
+				return nil
+			},
+		},
+		{
+			desc:    "success with viewer account",
```
Comment (kentakozuka) [inline pkg/account/api/api_key_test.go:496]: We need to check if the permission error is returned properly. So please use the viewer role for all cases and add an error case with the unassigned role.

**Review by kentakozuka — CHANGES_REQUESTED** [review]
Thanks! Left some comments. PTAL.

File: pkg/account/api/api_key_test.go (hunk 1 above)
Comment (kakcy) [reply pkg/account/api/api_key_test.go:448]: Indeed, in the account package test, AccountV2ByEnvironmentID() can be mocked within setup(), so there is no need to separate services for each test, so I will try to fix it.
I will look into other packages to see if they are possible.

File: pkg/account/api/api_key_test.go (hunk 2 above)
Comment (kakcy) [reply pkg/account/api/api_key_test.go:496]: @kentakozuka 

> We need to check if the permission error is returned properly. So please use the viewer role for all cases and add an error case with the unassigned role.

Please check whether the test case below matches your findings.

```
		{
			desc:    "errPermissionDenied",
			context: createContextWithDefaultToken(t, false),
			setup: func(s *AccountService) {
				s.accountStorage.(*accstoragemock.MockAccountStorage).EXPECT().GetAccountV2ByEnvironmentID(
					gomock.Any(), gomock.Any(), gomock.Any(),
				).Return(&domain.AccountV2{
					AccountV2: &accountproto.AccountV2{
						Email:            "email",
						OrganizationRole: accountproto.AccountV2_Role_Organization_UNASSIGNED,
						EnvironmentRoles: []*accountproto.AccountV2_EnvironmentRole{
							{
								EnvironmentId: "ns0",
								Role:          accountproto.AccountV2_Role_Environment_UNASSIGNED,
							},
						},
					},
				}, nil).AnyTimes()
			},
			req: &accountproto.GetAPIKeyRequest{Id: "id", EnvironmentNamespace: "ns0"},
			getExpectedErr: func(localizer locale.Localizer) error {
				return createError(localizer, statusPermissionDenied, localizer.MustLocalize(locale.PermissionDenied))
			},
		},
```

Comment (kentakozuka) [reply pkg/account/api/api_key_test.go:496]: Looks good!

**Review by kakcy — COMMENTED** [review]
> We can create the service in the test loop. plz remove the same line in the other tests as well.

In the account package test, I was able to generate the service in a loop, so I fixed it. But other It seems impossible to do so because the package requires a service to mock different permissions for each test case.

I'm soory, but because I changed the tests for multiple packages at the same time, there were a lot of differences.
From now on, we will issue a PR for each package.

**Review by kentakozuka — APPROVED** [review]
LGTM 👍

**Comment by kakcy** [comment]
Thank you!

**Pull request merged**